File: python/src/mpeg_o/cv_param.py
"""``CVParam`` — controlled-vocabulary parameter reference."""
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True, slots=True)
class CVParam:
    """A single controlled-vocabulary parameter reference.

    Immutable value class in the form
    ``(ontology_ref, accession, name, [value], [unit])``.

    Parameters
    ----------
    ontology_ref : str
        Ontology short name (``"MS"``, ``"NMR"``, ``"UO"``, ...).
    accession : str
        Ontology accession in ``<CV>:<id>`` form
        (e.g. ``"MS:1000515"``).
    name : str
        Human-readable label.
    value : str, default ""
        Optional free-form value. Use an empty string when the
        parameter is a pure assertion with no associated value.
    unit : str or None, default None
        Optional unit accession (typically a UO or MS term).

    Notes
    -----
    API status: Stable.

    Cross-language equivalents
    --------------------------
    Objective-C: ``MPGOCVParam`` · Java:
    ``com.dtwthalion.mpgo.CVParam``.
    """

    ontology_ref: str
    accession: str
    name: str
    value: str = ""
    unit: str | None = None

File: java/src/main/java/global/thalion/ttio/CVParam.java
/*
 * TTI-O Java Implementation
 * Copyright (C) 2026 DTW-Thalion
 * SPDX-License-Identifier: LGPL-3.0-or-later
 */
package global.thalion.ttio;

/**
 * A single controlled-vocabulary parameter reference. Immutable
 * value class.
 *
 * <p>In the form
 * {@code (ontologyRef, accession, name, [value], [unit])}. The
 * {@code accession} follows the {@code <CV>:<id>} form used by
 * PSI-MS (e.g. {@code "MS:1000515"}) or nmrCV
 * (e.g. {@code "NMR:1000002"}).</p>
 *
 * <p><b>API status:</b> Stable.</p>
 *
 * <p><b>Cross-language equivalents:</b> Objective-C
 * {@code TTIOCVParam}, Python
 * {@code ttio.cv_param.CVParam}.</p>
 *
 * @param ontologyRef Ontology short name ({@code "MS"}, {@code "NMR"}).
 * @param accession   Ontology accession in {@code <CV>:<id>} form.
 * @param name        Human-readable label.
 * @param value       Optional free-form value; empty string when none.
 * @param unit        Optional unit accession; {@code null} when none.
 *
 */
public record CVParam(
    String ontologyRef,
    String accession,
    String name,
    String value,
    String unit
) {}
